File: python/find_the_duplicate_number.py
from typing import List


class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        for i in range(len(nums)):
            to_swap_index = nums[i] - 1
            if i == to_swap_index:
                continue
            if nums[i] == nums[to_swap_index]:
                return nums[i]

            while nums[i] != i + 1:
                nums[i], nums[to_swap_index] = nums[to_swap_index], nums[i]

                if nums[i] == nums[to_swap_index]:
                    return nums[i]

                to_swap_index = nums[i] - 1


s = Solution()
print(s.findDuplicate([1, 3, 4, 2, 2]))
print(s.findDuplicate([3, 1, 3, 4, 2]))
print(s.findDuplicate([1, 1]))
print(s.findDuplicate([1, 1, 2]))
print(s.findDuplicate([2,1,2]))
print(s.findDuplicate([1, 3, 2, 4, 2]))
print(s.findDuplicate([8,7,1,10,17,15,18,11,16,9,19,12,5,14,3,4,2,13,18,18]))

File: scala/FindTheDuplicateNumber.scala
object FindTheDuplicateNumber extends App {
  def findDuplicate(nums: Array[Int]): Int = {
    val n = nums.length - 1
    var (l, r, c) = (1, n, n / 2)
    var answer = 1
    while (l <= r) {
      c = (l + r) / 2
      if(nums.count(_ <= c) > c) {
        answer = c
        r = c - 1
      } else {
        l = c + 1
      }
    }
    answer
  }

  println(findDuplicate(Array(1, 3, 4, 2, 2)))
  println(findDuplicate(Array(3, 1, 3, 4, 2)))
  println(findDuplicate(Array(2, 2, 2, 2, 2)))
  println(findDuplicate(Array(1, 1)))
  println(findDuplicate(Array(1,2,3,4,5,6,7,8,9,4)))

}
